### DISCUSSION TITLE
Custom Services and Eve App

### DISCUSSION THREAD

**Discussion by planetk** [HomeSpan/HomeSpan#206] [Q&A] [answered]
I am struggeling to make custom characteristics in Custom Services visible in the "Eve" app.

When I create an accessory with Custom Service "TestService" and Custom Characteristics "TestUINT8Char"(UINT8, PR,PW,EV) in the Apple Homekit Accessory Simulator, the accessory shows up in the Eve App. Further the Eve app shows a slider to manipulate the value of the characteristic.

When I define the exactly same Service and Characteristic (same UUID, name, rights, etc.) the Service, respective the characteristic does not show up in eve.

I'm using the homespan version 1.4.2. 

Characteristic and Service are defined as follows:

CUSTOM_CHAR(TestUINT8Char,  9EFEB3A9-4944-47F3-9CE4-EB281D639FDA, PR+EV+PW, UINT8, 120, 0, 255, false);

namespace Service {
  struct TestService : SpanService { TestService() : SpanService{"E415CC56-4CBE-40BC-B0C6-57C6E6EFD29B","TestService"}{
    req.push_back(&_CUSTOM_TestUINT8Char);
  }};
}

In the console HomeSpan Info and database look like shown below.

Does anybody hav a hint for me, what I'm doing wrong, or what s the difference between homespan and the homekit accessory simulator?

`
*** HomeSpan Info ***

➤ Accessory:  AID=1
   ➟ Service AccessoryInformation:  IID=1, UUID="3E"
      ⇨ Characteristic Name("TestAccessory"):  IID=2, UUID="23"
      ⇨ Characteristic Manufacturer("Tester"):  IID=3, UUID="20"
      ⇨ Characteristic SerialNumber("123456"):  IID=4, UUID="30"
      ⇨ Characteristic Model("TestModel"):  IID=5, UUID="21"
      ⇨ Characteristic FirmwareRevision("0.0.1"):  IID=6, UUID="52"
      ⇨ Characteristic Identify(0):  IID=7, UUID="14"
   ➟ Service HAPProtocolInformation:  IID=8, UUID="A2"
      ⇨ Characteristic Version("1.1.0"):  IID=9, UUID="37"
   ➟ Service TestService:  IID=10, UUID="E415CC56-4CBE-40BC-B0C6-57C6E6EFD29B"
      ⇨ Characteristic TestUINT8Char(120):  IID=11, Custom-UUID="9EFEB3A9-4944-47F3-9CE4-EB281D639FDA", Range=[0,255]

Range Check: No Warnings

Configured as Bridge: NO

Service                         UUID         AID  IID  Update  Loop  Button  Linked Services
------------------------------  ----  ----------  ---  ------  ----  ------  ---------------
AccessoryInformation              3E           1    1      NO    NO      NO  -
HAPProtocolInformation            A2           1    8      NO    NO      NO  -
TestService                     E415CC56-4CBE-40BC-B0C6-57C6E6EFD29B           1   10     YES   YES      NO  -

*** End Info ***

*** Attributes Database: size=805  configuration=18 ***

{
  "accessories":[
    {
      "aid":1,
      "services":[
        {
          "iid":1,
          "type":"3E",
          "characteristics":[
            {
              "iid":2,
              "type":"23",
              "value":"TestAccessory",
              "format":"string",
              "perms":[
                "pr"
              ]
            },
            {
              "iid":3,
              "type":"20",
              "value":"Tester",
              "format":"string",
              "perms":[
                "pr"
              ]
            },
            {
              "iid":4,
              "type":"30",
              "value":"123456",
              "format":"string",
              "perms":[
                "pr"
              ]
            },
            {
              "iid":5,
              "type":"21",
              "value":"TestModel",
              "format":"string",
              "perms":[
                "pr"
              ]
            },
            {
              "iid":6,
              "type":"52",
              "value":"0.0.1",
              "format":"string",
              "perms":[
                "pr"
              ]
            },
            {
              "iid":7,
              "type":"14",
              "format":"bool",
              "perms":[
                "pw"
              ]
            }
          ]
        },
        {
          "iid":8,
          "type":"A2",
          "characteristics":[
            {
              "iid":9,
              "type":"37",
              "value":"1.1.0",
              "format":"string",
              "perms":[
                "pr"
              ]
            }
          ]
        },
        {
          "iid":10,
          "type":"E415CC56-4CBE-40BC-B0C6-57C6E6EFD29B",
          "characteristics":[
            {
              "iid":11,
              "type":"9EFEB3A9-4944-47F3-9CE4-EB281D639FDA",
              "value":120,
              "format":"uint8",
              "perms":[
                "pr",
                "pw",
                "ev"
              ]
            }
          ]
        }
      ]
    }
  ]
}

*** End Database ***
`

**Comment by HomeSpan** [comment]
I've tested custom characteristics that extend a standard HomeKit service, such as adding "Voltage" to a LightBulb Service, and it's worked with Eve.  I've not tried adding a custom Service, though in principal this should work in HomeKit as well.  What functions do the custom UUIDs you have specified perform in Eve? I can try out in my setup to debug.

**Comment by planetk** [comment]
Hi,
thanx for your response. The specified custom uuids are just generated with the homekit accessory simulator. There is no specific function to this uuid.
My question is, why the same service and characteristic do show up when defined in Homespan, but are shown, when defined in the  homekit accessory simulator.
In the latter case the service shows up in Eve containing a slider for the UINT8 Characteristic allowing to set Values from 0-255.
This is more a general prove of concept, especially required to manage custom characteristics in actionsets (scenes) with eve .

**Comment by HomeSpan** [comment, marked as answer]
Puzzle solved!

HAP does not expect an application to consume a Characteristic or Service that it does not support.  For example, the Home App simply ignores all Services and Characteristics it does not recognize.  If you create a Characteristic that is recognized by Eve, but not by HomeSpan, that's fine as well since the Eve application knows what to do.

However, it seems that Eve also tries to "guess" how to render and display Services and Characteristics that it does not recognize at all.  I'm guessing that whoever developed this used the HomeKit simulator as the source of testing.  Through experimentation I found that for Custom *Services*, the HomeKit simulator seems to automatically add the Name Characteristic.  Though the Name Characteristic is generally optional, since the Home App and Eve know what to do based on the UUID of the Service, Eve seems to rely on this Characteristic being present to render a Custom Service.  So you must add `Characteristic::Name()` to a custom service unknown to Eve to allow Eve to display it.

In addition, you need to similarly provide a name for the Custom Characteristic itself.  This cannot be done with the Name Characteristic, since that's a separate Characteristic naming the Service.  Fortunately, HAP includes an optional "description" field as part of every Characteristic, and it seems to be this field that Eve is consuming.   In the production release of HomeSpan there is no method for adding a "description" to a Characteristic, but I just added `setDescription(const char *)` to the dev branch and tested that it works and allows Eve to render any Custom Characteristic it does not recognize (though you may need to use setRange() as well to ensure Eve knows the limits of the Characteristic value).

I'm not sure if there is any practical use to rendering a custom Characteristic in Eve, but at least it works if needed.

**Reply by planetk** [reply]
Wow! This is really good news! I‘m impressed, how you managed to find out about this. I‘ll give it a try soon. Thanks for your perserverance in solving this and for the detailed explanation!
